From 8f586d46d15a585f3554911f1c2fa8f57279196a Mon Sep 17 00:00:00 2001
From: Shreeram Ramamoorthy Swaminathan <shrerama@qti.qualcomm.com>
Date: Tue, 2 Jun 2026 23:17:40 +0530
Subject: [PATCH] [QMESH][RDK] WiFi8 Agent capabilities TLV for 11BN platforms

Added support for draft WiFi8 Agent capabilities TLV for 11BN platforms

Change-Id: Id76b28f9a74007bc7050bc74fd450864f913ba6b
Signed-off-by: Shreeram Ramamoorthy Swaminathan <shrerama@qti.qualcomm.com>
---
 .../slave/tasks/capability_reporting_task.cpp |  79 +++++
 .../slave/tasks/capability_reporting_task.h   |  14 +
 .../include/tlvf/wfa_map/eTlvTypeMap.h        |   3 +
 .../tlvf/wfa_map/tlvWifi8AgentCapabilities.h  | 131 ++++++++
 .../wfa_map/tlvWifi8AgentCapabilities.cpp     | 308 ++++++++++++++++++
 framework/tlvf/src/src/CmduMessageRx.cpp      |   4 +
 framework/tlvf/tlvf_conf.yaml                 |   1 +
 .../tlvf/yaml/tlvf/wfa_map/eTlvTypeMap.yaml   |   3 +
 .../wfa_map/tlvWifi8AgentCapabilities.yaml    |  81 +++++
 9 files changed, 624 insertions(+)
 create mode 100644 framework/tlvf/AutoGenerated/include/tlvf/wfa_map/tlvWifi8AgentCapabilities.h
 create mode 100644 framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvWifi8AgentCapabilities.cpp
 create mode 100644 framework/tlvf/yaml/tlvf/wfa_map/tlvWifi8AgentCapabilities.yaml

diff --git a/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp b/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
index e731e1458..c81b1a372 100644
--- a/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
+++ b/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
@@ -51,6 +51,9 @@
 #include <tlvf/wfa_map/tlvProfile2CacCapabilities.h>
 #include <tlvf/wfa_map/tlvProfile2MetricCollectionInterval.h>
 #include <tlvf/wfa_map/tlvWifi7AgentCapabilities.h>
+#if defined(IEEE80211_BN)
+#include <tlvf/wfa_map/tlvWifi8AgentCapabilities.h>
+#endif /* defined(IEEE80211_BN) */
 
 using namespace multi_vendor;
 
@@ -667,6 +670,12 @@ bool CapabilityReportingTask::prepare_ap_capability_message(bool early)
         LOG(ERROR) << "Error filling TLV_WIFI7_AGENT_CAPABILITIES";
         return false;
     }
+#if defined(IEEE80211_BN)
+    if (!add_wifi8_agent_capabilities_tlv(m_cmdu_tx)) {
+        LOG(ERROR) << "Error filling TLV_WIFI8_AGENT_CAPABILITIES";
+        return false;
+    }
+#endif /* defined(IEEE80211_BN) */
     if (!slave_thread::add_eht_operations_tlv(m_cmdu_tx)) {
         LOG(ERROR) << "Error filling TLV_EHT_OPERATIONS";
         return false;
@@ -1227,6 +1236,76 @@ bool CapabilityReportingTask::add_device_inventory_tlv()
     return true;
 }
 
+#if defined(IEEE80211_BN)
+bool CapabilityReportingTask::add_wifi8_agent_capabilities_tlv(
+    ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    auto db = AgentDB::get();
+
+    auto wifi8_agent_capabilities_tlv = cmdu_tx.addClass<wfa_map::tlvWifi8AgentCapabilities>();
+    if (!wifi8_agent_capabilities_tlv) {
+        LOG(ERROR) << "Error creating TLV_WIFI8_AGENT_CAPABILITIES";
+        return false;
+    }
+
+    for (auto radio : db->get_radios_list()) {
+        if (!radio) {
+            LOG(ERROR) << "radio does not exist in the db";
+            continue;
+        }
+
+        auto radio_wifi8_cap = wifi8_agent_capabilities_tlv->create_radio_wifi8_capabilities();
+        if (!radio_wifi8_cap) {
+            LOG(ERROR) << "create_radio_wifi8_capabilities() failed";
+            return false;
+        }
+
+        radio_wifi8_cap->ruid() = radio->front.iface_mac;
+
+        /*
+         * Hardcode all Wi-Fi 8 capability bits as supported/capable.
+         *
+         * flags1 (Byte 1):
+         *   Bit 7: smd_capable           - AP supports SMD
+         *   Bit 6: smd_dl_forwarding     - AP supports DL forwarding for SMD
+         *   Bit 5: per_smd_ptk_support   - per SMD PTK mode support
+         *   Bit 4: per_ap_ptk_support    - per AP PTK mode support
+         *   Bit 3: per_ap_mac_sap        - per AP Mac SAP support
+         *   Bit 2: per_smd_mac_sap       - per SMD Mac SAP support
+         *   Bit 1: coord_bf_support      - Co-ordinated BF support
+         *   Bit 0: coord_tdma_support    - Co-ordinated TDMA support
+         *
+         * flags2 (Byte 2):
+         *   Bit 7: npca_support          - Non-primary channel access support
+         *   Bits 6:0: reserved
+         *
+         * reserved (Byte 3): reserved for other Wi-Fi 8 features
+         */
+        radio_wifi8_cap->flags1().smd_capable         = 1;
+        radio_wifi8_cap->flags1().smd_dl_forwarding   = 1;
+        radio_wifi8_cap->flags1().per_smd_ptk_support = 1;
+        radio_wifi8_cap->flags1().per_ap_ptk_support  = 1;
+        radio_wifi8_cap->flags1().per_ap_mac_sap      = 1;
+        radio_wifi8_cap->flags1().per_smd_mac_sap     = 1;
+        radio_wifi8_cap->flags1().coord_bf_support    = 1;
+        radio_wifi8_cap->flags1().coord_tdma_support  = 1;
+
+        radio_wifi8_cap->flags2().npca_support = 1;
+        radio_wifi8_cap->flags2().reserved     = 0;
+
+        radio_wifi8_cap->reserved() = 0;
+
+        if (!wifi8_agent_capabilities_tlv->add_radio_wifi8_capabilities(radio_wifi8_cap)) {
+            LOG(ERROR) << "add_radio_wifi8_capabilities() failed in tlvWifi8AgentCapabilities";
+            return false;
+        }
+    }
+
+    LOG(INFO) << "Add WiFi 8 capabilities TLV";
+    return true;
+}
+#endif /* defined(IEEE80211_BN) */
+
 bool CapabilityReportingTask::add_profile2_ap_capability_tlv(ieee1905_1::CmduMessageTx &cmdu_tx)
 {
     auto profile2_ap_capability_tlv = cmdu_tx.addClass<wfa_map::tlvProfile2ApCapability>();
diff --git a/agent/src/beerocks/slave/tasks/capability_reporting_task.h b/agent/src/beerocks/slave/tasks/capability_reporting_task.h
index ff8c97a8f..62931f93f 100644
--- a/agent/src/beerocks/slave/tasks/capability_reporting_task.h
+++ b/agent/src/beerocks/slave/tasks/capability_reporting_task.h
@@ -142,6 +142,20 @@ private:
      */
     bool add_wifi7_agent_capabilities_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
 
+#if defined(IEEE80211_BN)
+    /**
+     * @brief Adds a new WiFi 8 Agent Capabilities TLV to given message.
+     *
+     * The TLV is defined in the IEEE 802.11bn (Wi-Fi 8) Multi-AP specification.
+     * All capability bits are hardcoded as supported/capable.
+     *
+     * @param[in,out] cmdu_tx CMDU message.
+     *
+     * @return True on success and false otherwise.
+     */
+    bool add_wifi8_agent_capabilities_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+#endif /* defined(IEEE80211_BN) */
+
 public:
     /* Note:
      * Profile-2 AP Capability TLV is being added by to the AutoConfiguration Message with M1, and
diff --git a/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/eTlvTypeMap.h b/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/eTlvTypeMap.h
index e56bf844d..aa428da4e 100644
--- a/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/eTlvTypeMap.h
+++ b/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/eTlvTypeMap.h
@@ -127,6 +127,7 @@ enum class eTlvTypeMap : uint8_t {
     TLV_AVAILABLE_SPECTRUM_INQUIRY_RESPONSE = 0xe9,
     TLV_RSN_PARAMETERS_CONFIGURATION = 0xeb,
     TLV_BSS_ADVANCED_CONFIGURATION = 0xec,
+    TLV_WIFI_8_AGENT_CAPABILITIES = 0xed,
     TLV_AP_WIFI_6_CAPABILITIES = 0xaa,
 };
 // Enum AutoPrint generated code snippet begining- DON'T EDIT!
@@ -239,6 +240,7 @@ static const char *eTlvTypeMap_str(eTlvTypeMap enum_value) {
     case eTlvTypeMap::TLV_AVAILABLE_SPECTRUM_INQUIRY_RESPONSE:      return "eTlvTypeMap::TLV_AVAILABLE_SPECTRUM_INQUIRY_RESPONSE";
     case eTlvTypeMap::TLV_RSN_PARAMETERS_CONFIGURATION:             return "eTlvTypeMap::TLV_RSN_PARAMETERS_CONFIGURATION";
     case eTlvTypeMap::TLV_BSS_ADVANCED_CONFIGURATION:               return "eTlvTypeMap::TLV_BSS_ADVANCED_CONFIGURATION";
+    case eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES:                return "eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES";
     case eTlvTypeMap::TLV_AP_WIFI_6_CAPABILITIES:                   return "eTlvTypeMap::TLV_AP_WIFI_6_CAPABILITIES";
     }
     static std::string out_str = std::to_string(int(enum_value));
@@ -358,6 +360,7 @@ public:
         case 0xe9:
         case 0xeb:
         case 0xec:
+        case 0xed:
         case 0xaa:
                 ret = true;
                 break;
diff --git a/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/tlvWifi8AgentCapabilities.h b/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/tlvWifi8AgentCapabilities.h
new file mode 100644
index 000000000..d14271afb
--- /dev/null
+++ b/framework/tlvf/AutoGenerated/include/tlvf/wfa_map/tlvWifi8AgentCapabilities.h
@@ -0,0 +1,131 @@
+///////////////////////////////////////
+// AUTO GENERATED FILE - DO NOT EDIT //
+///////////////////////////////////////
+
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TLVF_WFA_MAP_TLVWIFI8AGENTCAPABILITIES_H_
+#define _TLVF_WFA_MAP_TLVWIFI8AGENTCAPABILITIES_H_
+
+#include <cstddef>
+#include <stdint.h>
+#include <tlvf/swap.h>
+#include <string.h>
+#include <memory>
+#include <tlvf/BaseClass.h>
+#include <tlvf/ClassList.h>
+#include "tlvf/wfa_map/eTlvTypeMap.h"
+#include <tuple>
+#include <vector>
+#include "tlvf/common/sMacAddr.h"
+#include <asm/byteorder.h>
+
+namespace wfa_map {
+
+class cRadioWifi8Capabilities;
+
+class tlvWifi8AgentCapabilities : public BaseClass
+{
+    public:
+        tlvWifi8AgentCapabilities(uint8_t* buff, size_t buff_len, bool parse = false);
+        explicit tlvWifi8AgentCapabilities(std::shared_ptr<BaseClass> base, bool parse = false);
+        ~tlvWifi8AgentCapabilities();
+
+        const eTlvTypeMap& type();
+        const uint16_t& length();
+        uint8_t& num_radio();
+        std::tuple<bool, cRadioWifi8Capabilities&> radio_wifi8_capabilities(size_t idx);
+        std::shared_ptr<cRadioWifi8Capabilities> create_radio_wifi8_capabilities();
+        bool add_radio_wifi8_capabilities(std::shared_ptr<cRadioWifi8Capabilities> ptr);
+        void class_swap() override;
+        bool finalize() override;
+        static size_t get_initial_size();
+
+    private:
+        bool init();
+        eTlvTypeMap* m_type = nullptr;
+        uint16_t* m_length = nullptr;
+        uint8_t* m_num_radio = nullptr;
+        cRadioWifi8Capabilities* m_radio_wifi8_capabilities = nullptr;
+        size_t m_radio_wifi8_capabilities_idx__ = 0;
+        std::vector<std::shared_ptr<cRadioWifi8Capabilities>> m_radio_wifi8_capabilities_vector;
+        bool m_lock_allocation__ = false;
+        int m_lock_order_counter__ = 0;
+};
+
+class cRadioWifi8Capabilities : public BaseClass
+{
+    public:
+        cRadioWifi8Capabilities(uint8_t* buff, size_t buff_len, bool parse = false);
+        explicit cRadioWifi8Capabilities(std::shared_ptr<BaseClass> base, bool parse = false);
+        ~cRadioWifi8Capabilities();
+
+        typedef struct sWifi8Flags1 {
+            #if defined(__LITTLE_ENDIAN_BITFIELD)
+            uint8_t coord_tdma_support : 1;
+            uint8_t coord_bf_support : 1;
+            uint8_t per_smd_mac_sap : 1;
+            uint8_t per_ap_mac_sap : 1;
+            uint8_t per_ap_ptk_support : 1;
+            uint8_t per_smd_ptk_support : 1;
+            uint8_t smd_dl_forwarding : 1;
+            uint8_t smd_capable : 1;
+            #elif defined(__BIG_ENDIAN_BITFIELD)
+            uint8_t smd_capable : 1;
+            uint8_t smd_dl_forwarding : 1;
+            uint8_t per_smd_ptk_support : 1;
+            uint8_t per_ap_ptk_support : 1;
+            uint8_t per_ap_mac_sap : 1;
+            uint8_t per_smd_mac_sap : 1;
+            uint8_t coord_bf_support : 1;
+            uint8_t coord_tdma_support : 1;
+            #else
+            #error "Bitfield macros are not defined"
+            #endif
+            void struct_swap(){
+            }
+            void struct_init(){
+            }
+        } __attribute__((packed)) sWifi8Flags1;
+
+        typedef struct sWifi8Flags2 {
+            #if defined(__LITTLE_ENDIAN_BITFIELD)
+            uint8_t reserved : 7;
+            uint8_t npca_support : 1;
+            #elif defined(__BIG_ENDIAN_BITFIELD)
+            uint8_t npca_support : 1;
+            uint8_t reserved : 7;
+            #else
+            #error "Bitfield macros are not defined"
+            #endif
+            void struct_swap(){
+            }
+            void struct_init(){
+            }
+        } __attribute__((packed)) sWifi8Flags2;
+
+        sMacAddr& ruid();
+        sWifi8Flags1& flags1();
+        sWifi8Flags2& flags2();
+        uint8_t& reserved();
+        void class_swap() override;
+        bool finalize() override;
+        static size_t get_initial_size();
+
+    private:
+        bool init();
+        sMacAddr* m_ruid = nullptr;
+        sWifi8Flags1* m_flags1 = nullptr;
+        sWifi8Flags2* m_flags2 = nullptr;
+        uint8_t* m_reserved = nullptr;
+};
+
+}; // close namespace: wfa_map
+
+#endif //_TLVF/WFA_MAP_TLVWIFI8AGENTCAPABILITIES_H_
diff --git a/framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvWifi8AgentCapabilities.cpp b/framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvWifi8AgentCapabilities.cpp
new file mode 100644
index 000000000..996bd8cac
--- /dev/null
+++ b/framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvWifi8AgentCapabilities.cpp
@@ -0,0 +1,308 @@
+///////////////////////////////////////
+// AUTO GENERATED FILE - DO NOT EDIT //
+///////////////////////////////////////
+
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <tlvf/wfa_map/tlvWifi8AgentCapabilities.h>
+#include <tlvf/tlvflogging.h>
+
+using namespace wfa_map;
+
+tlvWifi8AgentCapabilities::tlvWifi8AgentCapabilities(uint8_t* buff, size_t buff_len, bool parse) :
+    BaseClass(buff, buff_len, parse) {
+    m_init_succeeded = init();
+}
+tlvWifi8AgentCapabilities::tlvWifi8AgentCapabilities(std::shared_ptr<BaseClass> base, bool parse) :
+BaseClass(base->getBuffPtr(), base->getBuffRemainingBytes(), parse){
+    m_init_succeeded = init();
+}
+tlvWifi8AgentCapabilities::~tlvWifi8AgentCapabilities() {
+}
+const eTlvTypeMap& tlvWifi8AgentCapabilities::type() {
+    return (const eTlvTypeMap&)(*m_type);
+}
+
+const uint16_t& tlvWifi8AgentCapabilities::length() {
+    return (const uint16_t&)(*m_length);
+}
+
+uint8_t& tlvWifi8AgentCapabilities::num_radio() {
+    return (uint8_t&)(*m_num_radio);
+}
+
+std::tuple<bool, cRadioWifi8Capabilities&> tlvWifi8AgentCapabilities::radio_wifi8_capabilities(size_t idx) {
+    bool ret_success = ( (m_radio_wifi8_capabilities_idx__ > 0) && (m_radio_wifi8_capabilities_idx__ > idx) );
+    size_t ret_idx = ret_success ? idx : 0;
+    if (!ret_success) {
+        TLVF_LOG(ERROR) << "Requested index is greater than the number of available entries";
+    }
+    return std::forward_as_tuple(ret_success, *(m_radio_wifi8_capabilities_vector[ret_idx]));
+}
+
+std::shared_ptr<cRadioWifi8Capabilities> tlvWifi8AgentCapabilities::create_radio_wifi8_capabilities() {
+    if (m_lock_order_counter__ > 0) {
+        TLVF_LOG(ERROR) << "Out of order allocation for variable length list radio_wifi8_capabilities, abort!";
+        return nullptr;
+    }
+    size_t len = cRadioWifi8Capabilities::get_initial_size();
+    if (m_lock_allocation__) {
+        TLVF_LOG(ERROR) << "Can't create new element before adding the previous one";
+        return nullptr;
+    }
+    if (getBuffRemainingBytes() < len) {
+        TLVF_LOG(ERROR) << "Not enough available space on buffer";
+        return nullptr;
+    }
+    m_lock_order_counter__ = 0;
+    m_lock_allocation__ = true;
+    uint8_t *src = (uint8_t *)m_radio_wifi8_capabilities;
+    if (m_radio_wifi8_capabilities_idx__ > 0) {
+        src = (uint8_t *)m_radio_wifi8_capabilities_vector[m_radio_wifi8_capabilities_idx__ - 1]->getBuffPtr();
+    }
+    if (!m_parse__) {
+        uint8_t *dst = src + len;
+        size_t move_length = getBuffRemainingBytes(src) - len;
+        std::copy_n(src, move_length, dst);
+    }
+    return std::make_shared<cRadioWifi8Capabilities>(src, getBuffRemainingBytes(src), m_parse__);
+}
+
+bool tlvWifi8AgentCapabilities::add_radio_wifi8_capabilities(std::shared_ptr<cRadioWifi8Capabilities> ptr) {
+    if (ptr == nullptr) {
+        TLVF_LOG(ERROR) << "Received entry is nullptr";
+        return false;
+    }
+    if (m_lock_allocation__ == false) {
+        TLVF_LOG(ERROR) << "No call to create_radio_wifi8_capabilities was called before add_radio_wifi8_capabilities";
+        return false;
+    }
+    uint8_t *src = (uint8_t *)m_radio_wifi8_capabilities;
+    if (m_radio_wifi8_capabilities_idx__ > 0) {
+        src = (uint8_t *)m_radio_wifi8_capabilities_vector[m_radio_wifi8_capabilities_idx__ - 1]->getBuffPtr();
+    }
+    if (ptr->getStartBuffPtr() != src) {
+        TLVF_LOG(ERROR) << "Received entry pointer is different than expected (expecting the same pointer returned from add method)";
+        return false;
+    }
+    if (ptr->getLen() > getBuffRemainingBytes(ptr->getStartBuffPtr())) {;
+        TLVF_LOG(ERROR) << "Not enough available space on buffer";
+        return false;
+    }
+    m_radio_wifi8_capabilities_idx__++;
+    if (!m_parse__) { (*m_num_radio)++; }
+    size_t len = ptr->getLen();
+    m_radio_wifi8_capabilities_vector.push_back(ptr);
+    if (!buffPtrIncrementSafe(len)) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << len << ") Failed!";
+        return false;
+    }
+    if(!m_parse__ && m_length){ (*m_length) += len; }
+    m_lock_allocation__ = false;
+    return true;
+}
+
+void tlvWifi8AgentCapabilities::class_swap()
+{
+    tlvf_swap(16, reinterpret_cast<uint8_t*>(m_length));
+    for (size_t i = 0; i < m_radio_wifi8_capabilities_idx__; i++){
+        std::get<1>(radio_wifi8_capabilities(i)).class_swap();
+    }
+}
+
+bool tlvWifi8AgentCapabilities::finalize()
+{
+    if (m_parse__) {
+        TLVF_LOG(DEBUG) << "finalize() called but m_parse__ is set";
+        return true;
+    }
+    if (m_finalized__) {
+        TLVF_LOG(DEBUG) << "finalize() called for already finalized class";
+        return true;
+    }
+    if (!isPostInitSucceeded()) {
+        TLVF_LOG(ERROR) << "post init check failed";
+        return false;
+    }
+    if (m_inner__) {
+        if (!m_inner__->finalize()) {
+            TLVF_LOG(ERROR) << "m_inner__->finalize() failed";
+            return false;
+        }
+        auto tailroom = m_inner__->getMessageBuffLength() - m_inner__->getMessageLength();
+        m_buff_ptr__ -= tailroom;
+        *m_length -= tailroom;
+    }
+    class_swap();
+    m_finalized__ = true;
+    return true;
+}
+
+size_t tlvWifi8AgentCapabilities::get_initial_size()
+{
+    size_t class_size = 0;
+    class_size += sizeof(eTlvTypeMap); // type
+    class_size += sizeof(uint16_t); // length
+    class_size += sizeof(uint8_t); // num_radio
+    return class_size;
+}
+
+bool tlvWifi8AgentCapabilities::init()
+{
+    if (getBuffRemainingBytes() < get_initial_size()) {
+        TLVF_LOG(ERROR) << "Not enough available space on buffer. Class init failed";
+        return false;
+    }
+    m_type = reinterpret_cast<eTlvTypeMap*>(m_buff_ptr__);
+    if (!m_parse__) *m_type = eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES;
+    if (!buffPtrIncrementSafe(sizeof(eTlvTypeMap))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(eTlvTypeMap) << ") Failed!";
+        return false;
+    }
+    m_length = reinterpret_cast<uint16_t*>(m_buff_ptr__);
+    if (!m_parse__) *m_length = 0;
+    if (!buffPtrIncrementSafe(sizeof(uint16_t))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint16_t) << ") Failed!";
+        return false;
+    }
+    m_num_radio = reinterpret_cast<uint8_t*>(m_buff_ptr__);
+    if (!m_parse__) *m_num_radio = 0;
+    if (!buffPtrIncrementSafe(sizeof(uint8_t))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint8_t) << ") Failed!";
+        return false;
+    }
+    if(m_length && !m_parse__){ (*m_length) += sizeof(uint8_t); }
+    m_radio_wifi8_capabilities = reinterpret_cast<cRadioWifi8Capabilities*>(m_buff_ptr__);
+    uint8_t num_radio = *m_num_radio;
+    m_radio_wifi8_capabilities_idx__ = 0;
+    for (size_t i = 0; i < num_radio; i++) {
+        auto radio_wifi8_capabilities = create_radio_wifi8_capabilities();
+        if (!radio_wifi8_capabilities || !radio_wifi8_capabilities->isInitialized()) {
+            TLVF_LOG(ERROR) << "create_radio_wifi8_capabilities() failed";
+            return false;
+        }
+        if (!add_radio_wifi8_capabilities(radio_wifi8_capabilities)) {
+            TLVF_LOG(ERROR) << "add_radio_wifi8_capabilities() failed";
+            return false;
+        }
+        // swap back since radio_wifi8_capabilities will be swapped as part of the whole class swap
+        radio_wifi8_capabilities->class_swap();
+    }
+    if (m_parse__) { class_swap(); }
+    if (m_parse__) {
+        if (*m_type != eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES) {
+            TLVF_LOG(ERROR) << "TLV type mismatch. Expected value: " << int(eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES) << ", received value: " << int(*m_type);
+            return false;
+        }
+    }
+    return true;
+}
+
+cRadioWifi8Capabilities::cRadioWifi8Capabilities(uint8_t* buff, size_t buff_len, bool parse) :
+    BaseClass(buff, buff_len, parse) {
+    m_init_succeeded = init();
+}
+cRadioWifi8Capabilities::cRadioWifi8Capabilities(std::shared_ptr<BaseClass> base, bool parse) :
+BaseClass(base->getBuffPtr(), base->getBuffRemainingBytes(), parse){
+    m_init_succeeded = init();
+}
+cRadioWifi8Capabilities::~cRadioWifi8Capabilities() {
+}
+sMacAddr& cRadioWifi8Capabilities::ruid() {
+    return (sMacAddr&)(*m_ruid);
+}
+
+cRadioWifi8Capabilities::sWifi8Flags1& cRadioWifi8Capabilities::flags1() {
+    return (sWifi8Flags1&)(*m_flags1);
+}
+
+cRadioWifi8Capabilities::sWifi8Flags2& cRadioWifi8Capabilities::flags2() {
+    return (sWifi8Flags2&)(*m_flags2);
+}
+
+uint8_t& cRadioWifi8Capabilities::reserved() {
+    return (uint8_t&)(*m_reserved);
+}
+
+void cRadioWifi8Capabilities::class_swap()
+{
+    m_ruid->struct_swap();
+    m_flags1->struct_swap();
+    m_flags2->struct_swap();
+}
+
+bool cRadioWifi8Capabilities::finalize()
+{
+    if (m_parse__) {
+        TLVF_LOG(DEBUG) << "finalize() called but m_parse__ is set";
+        return true;
+    }
+    if (m_finalized__) {
+        TLVF_LOG(DEBUG) << "finalize() called for already finalized class";
+        return true;
+    }
+    if (!isPostInitSucceeded()) {
+        TLVF_LOG(ERROR) << "post init check failed";
+        return false;
+    }
+    if (m_inner__) {
+        if (!m_inner__->finalize()) {
+            TLVF_LOG(ERROR) << "m_inner__->finalize() failed";
+            return false;
+        }
+        auto tailroom = m_inner__->getMessageBuffLength() - m_inner__->getMessageLength();
+        m_buff_ptr__ -= tailroom;
+    }
+    class_swap();
+    m_finalized__ = true;
+    return true;
+}
+
+size_t cRadioWifi8Capabilities::get_initial_size()
+{
+    size_t class_size = 0;
+    class_size += sizeof(sMacAddr); // ruid
+    class_size += sizeof(sWifi8Flags1); // flags1
+    class_size += sizeof(sWifi8Flags2); // flags2
+    class_size += sizeof(uint8_t); // reserved
+    return class_size;
+}
+
+bool cRadioWifi8Capabilities::init()
+{
+    if (getBuffRemainingBytes() < get_initial_size()) {
+        TLVF_LOG(ERROR) << "Not enough available space on buffer. Class init failed";
+        return false;
+    }
+    m_ruid = reinterpret_cast<sMacAddr*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sMacAddr))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sMacAddr) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_ruid->struct_init(); }
+    m_flags1 = reinterpret_cast<sWifi8Flags1*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sWifi8Flags1))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sWifi8Flags1) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_flags1->struct_init(); }
+    m_flags2 = reinterpret_cast<sWifi8Flags2*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sWifi8Flags2))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sWifi8Flags2) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_flags2->struct_init(); }
+    m_reserved = reinterpret_cast<uint8_t*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(uint8_t))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint8_t) << ") Failed!";
+        return false;
+    }
+    if (m_parse__) { class_swap(); }
+    return true;
+}
diff --git a/framework/tlvf/src/src/CmduMessageRx.cpp b/framework/tlvf/src/src/CmduMessageRx.cpp
index 545bdfc28..3bdf17c0c 100644
--- a/framework/tlvf/src/src/CmduMessageRx.cpp
+++ b/framework/tlvf/src/src/CmduMessageRx.cpp
@@ -147,6 +147,7 @@
 #include <tlvf/wfa_map/tlvVirtualBssDestruction.h>
 #include <tlvf/wfa_map/tlvVirtualBssEvent.h>
 #include <tlvf/wfa_map/tlvWifi7AgentCapabilities.h>
+#include <tlvf/wfa_map/tlvWifi8AgentCapabilities.h>
 
 using namespace ieee1905_1;
 
@@ -637,6 +638,9 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv(wfa_map::eTlvTypeMap tlv_
     case (wfa_map::eTlvTypeMap::TLV_BSS_ADVANCED_CONFIGURATION): {
         return msg.addClass<wfa_map::tlvBssAdvancedConfiguration>();
     }
+    case (wfa_map::eTlvTypeMap::TLV_WIFI_8_AGENT_CAPABILITIES): {
+        return msg.addClass<wfa_map::tlvWifi8AgentCapabilities>();
+    }
     }
     LOG(FATAL) << "Unknown TLV type: " << unsigned(tlv_type);
     return msg.addClass<tlvUnknown>();
diff --git a/framework/tlvf/tlvf_conf.yaml b/framework/tlvf/tlvf_conf.yaml
index a970ccba7..44d9ac117 100644
--- a/framework/tlvf/tlvf_conf.yaml
+++ b/framework/tlvf/tlvf_conf.yaml
@@ -127,6 +127,7 @@ include_yaml_path: {
     "tlvf/wfa_map/tlvSpatialReuseRequest.yaml",
     "tlvf/wfa_map/tlvSpatialReuseReport.yaml",
     "tlvf/wfa_map/tlvWifi7AgentCapabilities.yaml",
+    "tlvf/wfa_map/tlvWifi8AgentCapabilities.yaml",
     "tlvf/wfa_map/tlvAgentApMldConfiguration.yaml",
     "tlvf/wfa_map/tlvBackhaulStaMldConfiguration.yaml",
     "tlvf/wfa_map/tlvAssociatedStaMldConfigurationReport.yaml",
diff --git a/framework/tlvf/yaml/tlvf/wfa_map/eTlvTypeMap.yaml b/framework/tlvf/yaml/tlvf/wfa_map/eTlvTypeMap.yaml
index 1c77f5be2..9bd8329a6 100644
--- a/framework/tlvf/yaml/tlvf/wfa_map/eTlvTypeMap.yaml
+++ b/framework/tlvf/yaml/tlvf/wfa_map/eTlvTypeMap.yaml
@@ -114,6 +114,9 @@ eTlvTypeMap:
   TLV_RSN_PARAMETERS_CONFIGURATION: 0xEB
   TLV_BSS_ADVANCED_CONFIGURATION: 0xEC
 
+  # Wi-Fi 8 (IEEE 802.11bn) starts here
+  TLV_WIFI_8_AGENT_CAPABILITIES: 0xED
+
   # Profile 3 starts here
   TLV_AP_WIFI_6_CAPABILITIES: 0xAA
   # When adding values, make sure to also update CmduMessageRx::parseNextTlv !
diff --git a/framework/tlvf/yaml/tlvf/wfa_map/tlvWifi8AgentCapabilities.yaml b/framework/tlvf/yaml/tlvf/wfa_map/tlvWifi8AgentCapabilities.yaml
new file mode 100644
index 000000000..09eccd84c
--- /dev/null
+++ b/framework/tlvf/yaml/tlvf/wfa_map/tlvWifi8AgentCapabilities.yaml
@@ -0,0 +1,81 @@
+#
+---
+_namespace: wfa_map
+
+tlvWifi8AgentCapabilities:
+  _type: class
+  _is_tlv_class : True
+  type:
+    _type: eTlvTypeMap
+    _value_const: TLV_WIFI_8_AGENT_CAPABILITIES
+  length: uint16_t
+  num_radio:
+    _type: uint8_t
+    _length_var: True
+  radio_wifi8_capabilities:
+    _type: cRadioWifi8Capabilities
+    _length: [ num_radio ]
+
+cRadioWifi8Capabilities:
+  _type: class
+  ruid:
+    _type: sMacAddr
+  flags1: sWifi8Flags1
+  flags2: sWifi8Flags2
+  reserved:
+    _type: uint8_t
+
+sWifi8Flags1:
+  _type: struct
+  _bit_field: uint8_t
+  smd_capable:
+    _bit_range: [7,7]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates whether AP supports SMD
+  smd_dl_forwarding:
+    _bit_range: [6,6]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates whether AP supports DL forwarding for SMD
+  per_smd_ptk_support:
+    _bit_range: [5,5]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates the support for per SMD PTK mode by the Agent
+  per_ap_ptk_support:
+    _bit_range: [4,4]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates the support for per AP PTK mode by the Agent
+  per_ap_mac_sap:
+    _bit_range: [3,3]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates the support for per AP Mac SAP by the Agent
+  per_smd_mac_sap:
+    _bit_range: [2,2]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates the support for per SMD Mac SAP by the Agent
+  coord_bf_support:
+    _bit_range: [1,1]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates whether AP supports Co-ordinated BF
+  coord_tdma_support:
+    _bit_range: [0,0]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates whether AP supports Co-ordinated TDMA
+
+sWifi8Flags2:
+  _type: struct
+  _bit_field: uint8_t
+  npca_support:
+    _bit_range: [7,7]
+    _comment: |
+      0: Not supported
+      1: Supported - Indicates whether AP supports Non-primary channel access (NPCA)
+  reserved:
+    _bit_range: [6,0]
-- 
2.34.1

